quickwit-search/src/cluster_client.rs

```diff
@@ -0,0 +1,541 @@
+        let (request, mut client) = placed_request;
+        let (result_sender, result_receiver) = unbounded_channel();
+        let client_pool = self.client_pool.clone();
+        let retry_policy = LeafSearchStreamRetryPolicy {};
+        tokio::spawn(async move {
+            let results_stream = client.leaf_search_stream(request.clone()).await;
+            let forward_result =
+                forward_leaf_search_stream(results_stream, result_sender.clone(), true).await;
+            if retry_policy.should_retry(&request, forward_result.as_ref()) {
+                let retry_request = retry_policy.retry_request(&request, forward_result.as_ref());
+                let retry_client_opt = retry_client(&client_pool, &client, &retry_request).await;
+                // Propagates the error if we cannot get a new client and stops the task.
+                if let Err(error) = retry_client_opt {
+                    let _ = result_sender.send(Err(SearchError::from(error)));
+                    return;
+                }
+                let mut retry_client = retry_client_opt.unwrap();
+                debug!(
+                    "Leaf search stream response error. Retry once to execute {:?} with {:?}",
+                    retry_request, client
+                );
+                let retry_results_stream = retry_client.leaf_search_stream(retry_request).await;
+                // It is ok to ignore send errors, there is nothing else to do.
+                let _ =
+                    forward_leaf_search_stream(retry_results_stream, result_sender.clone(), false)
+                        .await;
+            }
+        });
+
+        UnboundedReceiverStream::new(result_receiver)
+    }
+}
+
+// Merge initial leaf search results with results obtained from a retry.
+fn merge_leaf_search_result(
+    result: Result<LeafSearchResult, SearchError>,
+    retry_result: Result<LeafSearchResult, SearchError>,
+) -> Result<LeafSearchResult, SearchError> {
+    match (result, retry_result) {
+        (Ok(mut response), Ok(mut retry_response)) => {
+            response
+                .partial_hits
+                .append(&mut retry_response.partial_hits);
+            let merged_response = LeafSearchResult {
+                num_hits: response.num_hits + retry_response.num_hits,
+                num_attempted_splits: response.num_attempted_splits
+                    + retry_response.num_attempted_splits,
+                failed_splits: retry_response.failed_splits,
+                partial_hits: response.partial_hits,
+            };
+            Ok(merged_response)
+        }
+        (Ok(response), Err(_)) => Ok(response),
+        (Err(_), Ok(retry_response)) => Ok(retry_response),
+        (Err(error), Err(_)) => Err(error),
+    }
+}
+
+async fn forward_leaf_search_stream(
+    mut stream: UnboundedReceiverStream<Result<LeafSearchStreamResult, SearchError>>,
+    sender: UnboundedSender<Result<LeafSearchStreamResult, SearchError>>,
```

Comment:
Should we make both the stream and the sender Bounded instead?
(If it is sensible, we can put that in a separate ticket.)

The point would be to add some nice back pressure.

Reply:
I would prefer to handle that later when we enable HTTP streaming.

Reply:
ah yes